timelock-corruption-detection/src/main/java/com/palantir/timelock/corruption/detection/LocalTimestampInvariantsVerifier.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.timelock.corruption.detection;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Predicates;
+import com.google.common.collect.Maps;
+import com.google.common.collect.SetMultimap;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.paxos.ImmutableNamespaceAndUseCase;
+import com.palantir.paxos.NamespaceAndUseCase;
+import com.palantir.paxos.PaxosValue;
+import com.palantir.timelock.history.HistoryQuerySequenceBounds;
+import com.palantir.timelock.history.PaxosLogHistoryProgressTracker;
+import com.palantir.timelock.history.models.LearnerUseCase;
+import com.palantir.timelock.history.sqlite.SqlitePaxosStateLogHistory;
+import com.palantir.timelock.history.util.UseCaseUtils;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.sql.DataSource;
+import one.util.streamex.StreamEx;
+
+public class LocalTimestampInvariantsVerifier {
+    @VisibleForTesting
+    static final int DELTA = 5;
+
+    private final SqlitePaxosStateLogHistory sqlitePaxosStateLogHistory;
```

Comment:
This won't work, need to fetch the next seq number > upper and use range [lower, next_seq_after_upper] for analysis.